File: code/cpp/ch13/solutions/challenge_04_sol.cpp
/*
 * Solution for Challenge 4: Fence Painting
 * Chapter 13: Bronze Battle Plan — Complete Search & Simulation
 *
 * APPROACH: Sort by start, merge overlapping intervals, sum lengths.
 * TIME:  O(N log N)
 * SPACE: O(1)
 */

#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

int solve(vector<vector<int>> fences) {
    if (fences.empty()) return 0;
    sort(fences.begin(), fences.end());
    int total = 0;
    int cs = fences[0][0], ce = fences[0][1];
    for (int i = 1; i < (int)fences.size(); i++) {
        if (fences[i][0] <= ce) {
            ce = max(ce, fences[i][1]);
        } else {
            total += ce - cs;
            cs = fences[i][0];
            ce = fences[i][1];
        }
    }
    total += ce - cs;
    return total;
}

// -- Do not change anything below this line --------------------------
int main() {
    int n;
    cin >> n;
    vector<vector<int>> fences(n, vector<int>(2));
    for (int i = 0; i < n; i++) cin >> fences[i][0] >> fences[i][1];
    cout << solve(fences) << endl;
    return 0;
}

File: code/python/ch13/solutions/challenge_04_sol.py
"""
Solution for Challenge 4: Fence Painting (USACO Bronze Style)
============================================
Chapter 13: Bronze Battle Plan — Complete Search & Simulation

APPROACH
--------
Sort intervals by start. Merge overlapping intervals.
Sum the lengths of merged intervals.

TIME COMPLEXITY:  O(N log N) — sorting
SPACE COMPLEXITY: O(1) — in-place
"""


def solve(fences: list[list[int]]) -> int:
    """Return total painted length (no double counting)."""
    if not fences:
        return 0
    fences.sort()
    total = 0
    cur_start, cur_end = fences[0]
    for start, end in fences[1:]:
        if start <= cur_end:
            cur_end = max(cur_end, end)
        else:
            total += cur_end - cur_start
            cur_start, cur_end = start, end
    total += cur_end - cur_start
    return total


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    n = int(input())
    fences = []
    for _ in range(n):
        s, e = map(int, input().split())
        fences.append([s, e])
    print(solve(fences))
